Non implemented methods for routes can optionally return a 405 status

xdc0 opened the pull request.

Commit:
Non implemented methods for routes can optionally return a 405 status instead 404.

Commit:
Non implemented methods for routes can optionally return a 405 status instead 404.

Commit:
Non implemented methods for routes can optionally return a 405 status instead 404.

xdc0 commented:
This is a PR for [Issue 2414](https://github.com/strongloop/express/issues/2414)

dougwilson commented:
Hi, thank you so much! I looked at it just a bit and was playing, and found this test that fails, but I would expect it to pass if you can add it to the suite and then update the PR so it passes :)

``` js
  it('should return 405 for /foo', function (done) {
    var app = express();

    app.get('/bar', function(){});

    app.route('/foo', { automatic405: true })
    .get(function (req, res) {
      res.send('get');
    });

    request(app)
    .post('/foo')
    .expect(405, done);
  });
```

Commit:
Improvements on automatic405 routing behavior

Commit:
Improvements on automatic405 routing behavior

Commit:
Improvements on automatic405 routing behavior

xdc0 commented:
Thanks for the feedback!

I made some improvements on the implementation based on this feedback and some stuff I noticed after some review:
1. app-wide option is now set through `app.set('automatic 405 routing')`
2. There was a bug where we were checking the last route inspected, not the last matching route.
3. If the route has not been defined at any point, we return a 404, even if we have 'automatic 405 routing' option enabled, 405 should be sent for recognized URIs anyway so this seems like the correct behavior.

xdc0 commented:
Any new feedback with the latest changes?

dougwilson commented:
Hi @chuym , no, not yet, I haven't had a time to play with the changes, but I read your description above when you sent it and everything sounding fine, so I didn't really have any further comments before trying it out as well :) I will try to tonight.

dougwilson commented:
So, it looks like the 405 stuff is hiding errors. I would expect the following test to pass, as it it should get a 500 response back, but is instead getting a 405:

``` js
  it('should 500', function(done) {
    var app = express();
    app.set('automatic 405 routing', true);

    app.use(function(req, res, next){
      next(new Error('boom!'))
    });

    app.get('/bar', function(){});

    request(app)
    .post('/bar')
    .expect(500, done);
  });
```

xdc0 commented:
That's an interesting test case. When that same test is ran without the 405 option, the request returns a 404, looks like it's a separate bug.

dougwilson commented:
> When that same test is ran without the 405 option, the request returns a 404, looks like it's a separate bug.

Then that would indicate a regression in your code? Running that exact test on `master` results in a success, because it returns a 500. I can always commit it if you're not sure.

dougwilson committed:
tests: add test for matfching route after error

dougwilson commented:
The test is committed to master as 412a22d150b1f218e4fe30d8ce8a91fd3aff19c3 and passing. You can easily rebase on top of that commit to know if you're accidentally introducing a regression or not on top of the test doing the same with with the 405 enabled.

xdc0 commented:
Oops, should have checked in master as well! Indeed it passes. Thanks for adding the test, I'll check what's wrong. Thanks for your prompt help

dougwilson commented:
It's no problem! I know a lot of people are looking forward to this feature :) Let me know if there is ever any specific help you need as well :)

Commit:
Merge branch '405-not-allowed' of github.com:chuym/express into 405-not-allowed

Commit:
Fixed problem where errors are being silently ignored when 405 routing option is enabled

Commit:
Fixed problem where errors are being silently ignored when 405 routing option is enabled

felixfbecker commented:
When will this get merged?!

dougwilson commented:
GitHub does not notify when people push changes to a PR. My last message was that the OP had to tell me when the fixes were complete. There was never a response, thus no one knew if there was any more activity.

dougwilson commented:
Hi @chuym , were those pushes after we last spoke the finished product, or are you still in progress on the fixes?

xdc0 commented:
Hello @dougwilson The issues that were previously discussed have been fixed, there's no more changes to be done here

jsmithe commented:
Will this request also populate the Allow header?

dougwilson commented:
Good catch, @jsmithe , it looks like this does not set any headers. @chuym please add the `Allow` header just like how the `OPTIONS` auto response works :)

felixfbecker commented:
Yeah Allow header is a must-have.

Commit:
Merge branch '405-not-allowed' of github.com:chuym/express into 405-not-allowed

Commit:
Adding 'Allowed' in header when returning 405

Commit:
Passing methods as strings instead array.

xdc0 commented:
@dougwilson @jsmithe Added 'Allow' header when returning a 405 response as it should. Please let me know if there is any other issue.

jsmithe commented:
405 is returned even when the method is OPTIONS.  So it is preventing the automatic options response from occurring.   Was this intended?  If not, would it be possible to work it back in?

felixfbecker commented:
Lol OPTIONS should always be allowed and handled automatically, shouldn't it?

jsmithe commented:
If that is the case should there be a test for it in the base?

Commit:
options default handler is considered when doing a 405 response

xdc0 commented:
@jsmithe Indeed, express does have a default handler when the `OPTIONS` method is not defined explicitly, the 405 function is not considering this default handler when `OPTIONS` is not explicitly defined with the route. I have fixed this and added tests that covers this default handler for the `OPTIONS` method

felixfbecker commented:
nice

jsmithe commented:
All routes are looking good in my project with this cut.  Thanks for the effort on this addition.

sashahilton00 commented:
@dougwilson bumping this, think it was forgotten :)

okonon commented:
@dougwilson are there any plans to merge this PR?

dougwilson commented:
Hmm, looks like there are conflicts that even github cannot resolve that would need to happen. This is so old, it looks like before the router was removed from here and exported as the "router" module.

There isn't any current plan to land, but I can add it to my backlog to go over this PR and write up the state of it and what would need to be done to land.

okonon commented:
@dougwilson understood! thanks for prompt response

dougwilson commented:
No problem! And just to be clear, at this point I'm not asking anything of the OP, because it may be straight forward and I can just resolve the conflicts & land. A new Express 4.16 is upcoming, and so looking for things to land in it regardless.

okonon commented:
Sounds good! thanks

xdc0 commented:
My apologies for not following up. At this point this is very likely not the way to solve the problem, or maybe not a problem anymore.
